### PR TITLE
[#6519] Add more checkstyle rules

### PR DESCRIPTION
**Pull request by amrut-prabhu**
Part of #6519 

**Outline of Solution**

More information on each of the checks [here](http://checkstyle.sourceforge.net/checks.html).
Added the following checkstyle rules and fixed any resultant errors:

- `ArrayTrailingComma`
- `FallThrough`
- `IllegalThrows`
- `MutableException`
- `OverloadMethodsDeclarationOrder `
- `IllegalType`
- `InterfaceIsType`
- `LambdaParameterName`
- `InnerTypeLast`

Upgraded Checkstyle from version `8.10.1` to `8.16` (latest version available in IntelliJ)

### CODE REVIEW

File: static-analysis/teammates-checkstyle.xml [hunk 1]
```diff
@@ -173,6 +173,27 @@
...
+    </module>
+    <module name="InterfaceIsType"/>
+    <module name="MutableException"/>
+    <module name="OverloadMethodsDeclarationOrder"/>
+    <module name="SummaryJavadocCheck"/>
+    <module name="JavadocStyle">
+      <property name="scope" value="package"/>
+      <property name="checkEmptyJavadoc" value="true"/>
+    </module>
+    <module name="ImportOrder">
```
Comment (wkurniawan07) [inline static-analysis/teammates-checkstyle.xml]: There are already configurations for this and `JavadocStyle`. Let's not be hasty.

File: static-analysis/teammates-checkstyle.xml [hunk 2]
```diff
@@ -173,6 +173,27 @@
       <property name="onCommentFormat" value="CHECKSTYLE\.ON:([A-Za-z\|]+)"/>
       <property name="checkFormat" value="$1"/>
     </module>
+    <module name="FallThrough"/>
+    <module name="ArrayTrailingComma"/>
+    <module name="IllegalThrows">
+      <property name="ignoredMethodNames" value="execute"/>
+      <!-- Remove execute value from this property after upgrading to JUnit 5 (see teammates.test.cases.Executable) -->
```
Comment (wkurniawan07) [inline static-analysis/teammates-checkstyle.xml]: If this is what you want, then you should put `CHECKSTYLE.OFF` around that method only, not suppressing for all methods named `execute`.

**Comment by amrut-prabhu** [comment]
Ready for review

**Comment by wkurniawan07** [comment]
@amrut-prabhu any reason not to include `IllegalType`, `InterfaceIsType`, `LambdaParameterName`?

**Comment by amrut-prabhu** [comment]
@wkurniawan07  
`IllegalType` causes 58 violations using the default settings due to the usage of HashMap, TreeMap etc. So, if we choose to add it, should we change the instances of these data types? Otherwise, what classes should be defined as illegal instead?

No specific reason for omitting `InterfaceIsType`.

As for `LabmdaParameterName`, I thought that it wasn't mentioned in our coding standard. After looking through it again, I realise that it is covered under `Variable names must be in camelCase.`  So, I'll proceed to add it.

Anything else needed?

Also, on a side note, just wanted to confirm that I need to create an issue before updating `teammates-checkstyle.xml` to comply with our spacing convention for XML (`4`,`4`,`8`)?

**Comment by wkurniawan07** [comment]
@amrut-prabhu 

You don't seem to understand the rule `IllegalType` well enough. It has something to do with [loose coupling](https://stackoverflow.com/questions/27868220/java-loose-coupling-avoid-using-implemnetation-types-like-arraylist-use-the-i), which is encouraged.

> updating `teammates-checkstyle.xml` to comply with our spacing convention for XML (`4`,`4`,`8`)?

How exactly is the current ruleset not enforcing that?

**Comment by amrut-prabhu** [comment]
Oh, okay. Thanks for the link. I'll proceed to make the necessary changes.

The [`teammates-checkstyle.xml`](https://github.com/TEAMMATES/teammates/blob/master/static-analysis/teammates-checkstyle.xml) file currently has spacing `2`,`2`,`4` compared to something like [`teammates-pmd.xml`](https://github.com/TEAMMATES/teammates/blob/master/static-analysis/teammates-pmd.xml).
Even IntelliJ pops up a message when opening the file: `This file is indented with 2 spaces instead of 4 spaces configured for the project`. 
Am I mistaken?

**Comment by wkurniawan07** [comment]
Ah I see, I misunderstood your comment, thinking that CS failed to enforce the indentation of Java files.
Feel free to correct the indentation of that XML file here in this PR.

**Pull request merged**